onnx/defs/shape_inference.h

```diff
@@ -75,15 +75,16 @@ struct InferenceContext {
 };
 
 struct DataPropagationContext {
+  virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
   virtual size_t getNumInputs() const = 0;
   virtual const TypeProto* getInputType(size_t index) const = 0;
   virtual size_t getNumOutputs() const = 0;
   virtual TypeProto* getOutputType(size_t index) = 0;
   virtual std::string getOutputName(size_t index) const = 0;
   virtual ~DataPropagationContext() {}
   virtual void addGeneratedShapeData(size_t index, TensorShapeProto&& tp) = 0;
-  virtual const TensorShapeProto* getGeneratedShapeData(size_t index) const = 0;
-  virtual const TensorShapeProto* getGeneratedShapeDataFromName(std::string name) const = 0;
+  virtual const TensorShapeProto* getInputGeneratedShapeData(size_t index) const = 0;
```

Comment:
nit: I still think it is better to replace ```getInputGeneratedShapeData``` by ```getInputData```. It is shorter, and more accurate. I am afraid the ```Shape``` part will likely confuse some developers. It is *not* related to the shape of the input ... it is actually information about the value of the input, (which may be used subsequently to infer the shape of some other tensor).

Reply:
Similarly, better to change ```getOutputGeneratedShapeData``` to ```getOutputData``` and ```addGeneratedShapeData``` to ```setOutputData```.

Reply:
agree with all the 3 recommendations

Reply:
Another suggestion is to add a ```typedef TensorShapeProto PartialInputData``` or ```typedef TensorShapeProto SymbolicInputData``` and use those new type names in this interface. We are reusing TensorShapeProto to avoid defining a new data-structure to store either constant-values or unknown-values. E.g., we could imagine a more sophisticated representation in the future.

Reply:
why is getOutputData required? When will this be used? You only setoutput data ... the next node calls getInputdata to fetch this data right?

Reply:
Just simplified the function name. GetOutputData is just for testing and won't be used in the propagation.

Yes it's still doable for testing without getOutputData... I have removed that function and add one more node behind the tested node to get the input data to verify.

Reply:
Originally InferenceContext has `getInputData` function to get "TensorProto". I think DataPropagationContext will need this function for future propagation functions as well... To sync the naming convention, I will make `getInputData` to get "TensorProto", `getInputShapeData"`to get the produced "TensorShapeProto", and `addOutputShapeData` to propagate the "TensorShapeProto". This PR is ready for review again @gramalingam @askhade. Thank you!